Lib/extractor/formats/opentype.py

```diff
@@ -1046,3 +1049,44 @@ def _extractOpenTypeKerningFromKern(source):
         # there are no minimum values.
         kerning.update(subtable.kernTable)
     return kerning
+
+# -----
+# Color
+# -----
+
+COLOR_LAYERS_KEY = "com.github.googlei18n.ufo2ft.colorLayers"
+COLOR_PALETTES_KEY = "com.github.googlei18n.ufo2ft.colorPalettes"
+COLOR_LAYER_MAPPING_KEY = "com.github.googlei18n.ufo2ft.colorLayerMapping"
+UFO2FT_FILTER_KEY = "com.github.googlei18n.ufo2ft.filters"
+def extractColorsFromOpenType(
+    source,
+    destination,
+):
+    if not "CPAL" in source or not "COLR" in source: return
+    destination.lib[UFO2FT_FILTER_KEY] = [{"name": "Explode Color Layer Glyphs", "pre": True}]
+    # UFO format only supports one palette. Others ignored.
```

Comment:
that's not true, colorPalettes key can contain multiple palette

it's not "UFO format", btw. This is currently a private set of keys only defined by ufo2ft, not made public in the UFO spec as of yet

Reply:
Uh, actually it's both. I added certain keys to support building in `fontmake`, but those keys aren't strictly necessary. If you go just by `layercontents.plist` and into each `glyphs` directory's `layerinfo.plist` you can easily build a color font from that. I assume that @justvanrossum didn't implement it this way, but instead did it by preprocessing the glyphs, to save time. Once @simoncozens works more on fonttools-rs I'll probably add COLR/CPAL support without these keys.

Reply:
But there's the rub: there's only a single `layercontents.plist`. You can't have more than one for different palettes (maybe something to add to UFO4?). Sure, these hacky keys support more than one palette, but I added the keys as an afterthought after I already made sure that the UFO generated correctly.

Reply:
we _deliberately_ chose not to use the layerinfo.plist color for this, since that one is traditionally used for GUI annotation purposes, in fact it's called `markColor` in defcon API for that reason.

> these hacky keys

I can accept that you say they are not documented properly, but they are not "hacky" keys. They are the ufo2ft's way to build COLR/CPAL fonts.

Reply:
I have no idea why it'd be expected that the layer color stops at the GUI, and I don't see that in the spec either?

Reply:
https://github.com/googlefonts/ufo2ft/pull/359#issuecomment-579326524

Reply:
Well, I disagree with the robofont developers' reading of the spec and my tools don't interpret it that way, so in my view adding extra keys _is_ hacky when there's already a standard way to color glyphs.

As @justvanrossum even wrote in that thread:

> The wording in the UFO spec suggests the layer color attribute could be used for color fonts, but doesn't say so explicitly.

So, it's not wrong to use it that way.

Multiple CPAL palettes are very seldom used…we could perhaps extend `layerinfo.plist` to refer to multiple colors to add multiple palettes as well in UFO4.

Reply:
The spec is ambiguous here because UFO3 is from 2012, dating before public color font specifications, the understood intent for `layerinfo.plist` is GUI. I'll try to get to some spec clarification on that this week, sorry for the confusion. For color fonts right now, the best way to do this is with the `ufo2ft` keys. If they seem useful, we can promote them to documented public keys, and then look at a way of doing all the various color font formats in UFO4.

Reply:
Unfortunately MFEKglif is already using `color` with the meaning I said here, it even supports exporting `layerinfo.plist` and `layercontents.plist`…

Reply:
you're free to interpret it as you like, what I'm saying is the spec is vague in that regard and doesn't fully convey the flexibility of COLR/CPAL and the exclusion of the layerinfo color was deliberate when Just and I discussed about the ufo2ft implementation. 

> If they seem useful, we can promote them to documented public keys

+1 (though I don't have much availability myself currently)

Reply:
Alright, I will leave the UFO3 spec be in the ambitious state. One is free to interpret it as they want, knowing that it may be for color fonts in one context, GUI in another. COLR/CPAL needs a good look for UFO4 (again, they didn't exist when UFO3 was specified, so right now everything is at some level of 'hacky').

@anthrotype, I'll see if I can slot that in this week

Reply:
OK, I'd say that worked out then. _As to this PR_, I don't mind writing multiple palettes to the appropriate key if they exist.

Reply:
> Unfortunately MFEKglif is already using color with the meaning I said here, it even supports exporting layerinfo.plist and layercontents.plist

I'm not sure it makes sense to say, "MFEKglif is already using this", given it has been available for zero years so far. I think it makes more sense to say "MFEKglif began using this but clearly the spec was ambiguous and getting some clarification on the 9 year old spec would be ideal"

Reply:
It makes sense to say it because a lot of work went into getting it to its current state based on the current reading, and more work would be needed to change it to a different reading. Also, the project is one year old in September, though the color layers are only a few months old, true. And only a few days old if you are only going by releases and not development branch activity.

I put more stock in how much effort something would take to change than how long it's been a certain way, though.

Reply:
@benkiel This probably isn't the place, but what about instead of promoting the keys, we add a disambiguating key? Something like `layer_colors_exported`, receiving a boolean? I can easily support that.

I just see it as duplicated data when most of the time users are going to want the GUI color and exported color to equal one another, in a world of color fonts.

As far as the multiple palettes go, this is going to have to go into UFO4. I'd naively solve it by making the layer `color` an array of strings instead of just a string. If it's missing from one layer and not another, then the color is equal across both palettes, and the layer with the most colors dictates how many palettes the font has. What do you think?

Reply:
I use colors a lot for grouping glyphs in the GUI, even in COLR/CPAL fonts, so making layer colors work also as CPAL colors would be the wrong thing for my fonts, since the two have different uses. It is better to keep the current UFO layer color as they have been understood by almost all implementations so far, and use a different structure for color font layers. Glyphs also uses distinct structures for GUI colors and CPAL palettes.

Reply:
This isn't a well-tread part of the spec, but since UFO4 will likely support COLRv1, this GUI color separation concern makes me think of the fact that COLRv1 has a flags array with masks for `USABLE_WITH_LIGHT_BACKGROUND` (0b1) and `USABLE_WITH_DARK_BACKGROUND` (0b10). (_Cf._ [Palette Type Array](https://docs.microsoft.com/en-us/typography/opentype/spec/cpal#palette-type-array), CPAL, OT Spec.) Since flags equal to 0b0 are meaningless (not usable at all?), that could be interpreted as, well, not usable at all; only a GUI coloring; not exported.

So if we're already saying that the `color` should take an array of strings, well, there'd also be a need to mark flags for each palette. Maybe, rather than an array of strings, it should be a structure similar to `[(flags, color), ...]`. That makes the most sense to me, and has the benefit of making the _default_ the current behavior, 0b0, not usable for any export.

Reply:
Or not? Do we think that the flags are useless and not worth supporting? I think they'll end up used eventually, emoji for use with dark themes seems to be the intent.